task/STC-13 Adjusted Routing and Logging

jcoro opened the pull request:
I changed the file upload configuration from `web.xml` to Java.  I also configured the file upload directory name in `pom.xml` with the value defined locally (in `settings.xml`).

jcoro committed:
[STC-13] Added functionality for users to upload, edit, and delete Datasets as well as the DataFiles they contain.

jcoro committed:
[STC-13] Updated functionality for users to manage Datasets as well as the DataFiles they contain.

jcoro committed:
[STC-13] Cleaned up files for Dataset and DataFile functionality.

jcoro committed:
[STC-13] Added pagination and a CustomPermissionEvaluator for securely updating and deleting Datasets and DataFiles

jcoro committed:
Merge branch 'develop' of https://github.com/diging/scriptoocloud into task/STC-13

 Conflicts:
	scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/SecurityContext.java
	scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/exceptions/DataFileStorageException.java
	scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/exceptions/DatasetStorageException.java
	scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/IFileSystemService.java
	scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java
	scriptoocloud/src/main/resources/config.properties

jcoro committed:
[STC-13] Adjusted Exception handling.

jcoro committed:
[STC-13] Changed file upload configuration from XML to Java.

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/CustomPermissionEvaluator.java:
```diff
@@ -0,0 +1,51 @@
+        if ((auth == null) || (targetType == null) || !(permission instanceof String)) {
+            return false;
+        }
+        if (targetType.equals(edu.asu.diging.scriptoocloud.core.model.impl.Dataset.class.getSimpleName())) {
```
Is there a reason you use the fully qualified name here instead of just `Dataset.class`?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/SecurityContext.java:
```diff
@@ -37,7 +37,8 @@ protected void configure(HttpSecurity http) throws Exception {
                 // The rest of the our application is protected.
-                .antMatchers( "/users/**", "/admin/**").hasRole("ADMIN")
+                .antMatchers("/users/**", "/admin/**").hasRole("ADMIN")
+                .antMatchers("/datasets/**").hasRole("USER")
```
I think I made this comment in another story before (maybe it's in one of the other branches?), but I would choose one prefix that is accessible by users with role `USER` and then use that for all parts of the application that require the user role (e.g. `/auth` or `/secured` or something). Your urls would then look like `/auth/datasets/...`, `/auth/projects/...`, etc. It's less error prone and doesn't require changes to the SecurityContext every time a new functionality is added.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/DataFile.java:
```diff
@@ -0,0 +1,61 @@
+import java.time.OffsetDateTime;
+
+@Entity
+public class DataFile {
```
needs javadoc

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/Dataset.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Set;
+
+@Entity
+public class Dataset implements IDataset {
```
needs javadoc

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java:
```diff
@@ -0,0 +1,91 @@
+            fileSystemService.createFileInDirectory(username, datasetId, filename, bytes);
+        } catch (FileSystemStorageException e) {
+            throw new DataFileStorageException("File could not be saved in the file system", e);
+        }
```
I would first do the database work and then save the file. Otherwise, if something on the database level goes wrong, the database transaction will be rolled back but you will still have the file lying around.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java:
```diff
@@ -0,0 +1,91 @@
+            datasetRepository.save(dataset.get());
+        } else {
+            throw new DataFileStorageException("Dataset not found");
+        }
```
I would return the created datafile here in case its ID is needed.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java:
```diff
@@ -0,0 +1,91 @@
+    @Override
+    public void createFile(byte[] bytes, String datasetId, String username, String filename, String type) throws DataFileStorageException {
+        try {
+            fileSystemService.createFileInDirectory(username, datasetId, filename, bytes);
```
Is `filename` the filename of the uploaded file? What if a user uploads two files with the same filename? Will the first one overridden by the second one? I would use the id of the datafile object as filename and then store the original filename in the db.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetService.java:
```diff
@@ -0,0 +1,192 @@
+                    user.getUsername());
+        } catch (FileSystemStorageException e) {
+            throw new DatasetStorageException("An IOException prevented the Dataset from being created", e);
+        }
```
I would return the dataset in case the caller needs its ID.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetService.java:
```diff
@@ -0,0 +1,192 @@
+        if (dataset.isPresent()) {
+            dataFileRepository.deleteAllByDatasetId(id);
+        }
+        datasetRepository.deleteById(id);
```
see above, I would first delete the data from the db then from the filesystem

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -32,10 +31,10 @@ public FileSystemService() {
      */
     @Override
-    public void addDatasetDirectories(String id, String username) throws DatasetStorageException {
+    public void addDatasetDirectories(String id, String username) throws FileSystemStorageException {
```
this doesn't necessarily have to be a dataset directory, right? This method would create a new directory in a user's folder no matter what it is used for. I would rename the method.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -45,52 +44,51 @@ public void addDatasetDirectories(String id, String username) throws DatasetStor
      */
     @Override
-    public void deleteDatasetDirectories(Long id, String username) throws SecurityException {
+    public void deleteDatasetDirectories(Long id, String username) throws FileSystemStorageException {
```
Similar to above, this doesn't necessarily have to be a dataset directory, right? This method would delete a directory in a user's folder no matter what it is used for. I would not pass a long but a string (so you can use it more easily later for other things) and rename the method.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddDatasetController.java:
```diff
@@ -0,0 +1,75 @@
+        try {
+            iDatasetService.createDataset(datasetForm.getName(), user);
+        } catch (DatasetStorageException e) {
+            Page<Dataset> dbDatasets = iDatasetService.findDatasets(PageRequest.of(currentPage - 1, pageSize), user);
```
you should log the exception so there is a record of it

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/AddDataFileController.java:
```diff
@@ -0,0 +1,86 @@
+            redirectAttributes.addFlashAttribute("errorMessage", "There was an error reading the file");
+        } catch (DataFileStorageException e) {
+            redirectAttributes.addFlashAttribute("errorMessage", "There was an error saving the file");
+        }
```
you should log the exceptions so there is a record of them

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DeleteDataFileController.java:
```diff
@@ -0,0 +1,48 @@
+            redirectAttributes.addFlashAttribute("errorMessage", "Dataset not found");
+        } catch (DataFileNotFoundException e) {
+            redirectAttributes.addFlashAttribute("errorMessage", "DataFile not found");
+        }
```
log exceptions

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DeleteDatasetController.java:
```diff
@@ -0,0 +1,41 @@
+            iDatasetService.deleteDataset(datasetId, username);
+        } catch (DatasetStorageException e) {
+            model.addAttribute("ErrorMessage", "Error: Could not delete Dataset");
+            return "redirect:/datasets/list";
```
log exceptions

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DeleteGitRepositoryController.java:
```diff
@@ -16,8 +18,12 @@
+        try{
+            gitRepositoryManager.deleteRepository(id);
+        } catch (FileSystemStorageException e){
+            redirectAttributes.addAttribute("formResponse","Error deleting repository");
```
log exception

jdamerow commented on scriptoocloud/src/main/resources/config.properties:
```diff
@@ -4,3 +4,5 @@ db.username=${db.user}
 git.repositories.path = ${git.repositories.path}
 digingUrl = http://diging.asu.edu/index.html
+rootUploadLocation = ${uploadDirectory}
+pageSize = 5
```
the default page size should probably be more like 25 or something? This is for datafiles, right?

jcoro replied:
I bumped the `pageSize` up to 25 as requested.  As it stands now, this `pageSize` is for both the `Dataset` list and the `DataFile` list, but if that needs to be adjusted, it's easy enough to do.  I started at 5 just for ease of manual testing.

jdamerow commented on scriptoocloud/src/main/webapp/WEB-INF/views/datasets/details.html:
```diff
@@ -0,0 +1,94 @@
+                <div class="alert alert-danger"
+                     th:if="${not #strings.isEmpty(errorMessage)}">
+                    <p th:text="${errorMessage}">
+                </div>
```
what are these used for? for general error messages? I would put them in your layout so they are available on every page.

jcoro replied:
These error and success messages are specific to `Dataset` operations on the `details.html` page.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/EditDatasetController.java:
```diff
@@ -0,0 +1,76 @@
+            }
+            model.addAttribute("dataset", new DatasetForm());
+            model.addAttribute("datasetEditForm", datasetEditForm);
+            model.addAttribute("errorIndex", datasetEditForm.getIndex());
```
I would not use an index to find out which dataset was edited but its id. What if someone else added another dataset and the index isn't the same anymore because of that after the user saves their edits?

jdamerow reviewed (changes requested):
- please reopen this pr again and don't create a new one, so that the comments are available.
- and for future stories, since STC-13 is a story, the branch name should be `story/STC-13` not `task/...`.

jcoro committed:
[STC-13] Adjusted routing and logging.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java:
```diff
@@ -0,0 +1,105 @@
+            dataset.get().getFiles().remove(newFile);
+            dataFileRepository.deleteById(dataFileId);
+            datasetRepository.save(dataset.get());
+            throw new DataFileStorageException("File could not be saved in the file system", e);
```
if you use `@Transactional(rollbackFor=Exception.class)`  and just throw the exception here, you should not have to revert these changes yourself. Spring will make sure to rollback your changes. If you do it by hand, this has to be kept in mind by future developers if they make any other changes to the data in the database.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetService.java:
```diff
@@ -0,0 +1,199 @@
+ * The Service class associated with user Datasets.  The class handles the association between
+ * Datasets and DataFiles they contain.
+ */
+@Transactional
```
same as above use the `rollbackFor` attribute to rollback any database changes in case of an exception

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -28,69 +27,72 @@ public FileSystemService() {
-     * @param datasetId The id of the dataset.
+     * @param username The username of the owner of the dataset.
+     * @param type     The directory type (e.g., Dataset, etc.)
+     * @param id       The id name.
```
remove `name`? Also, this method will still work if id is `null`, right? The javadoc should say that. And last, you probably want to move the javadoc to the interface otherwise the autocomplete menu in Eclipse (and I presume IntelliJ as well) won't show it when the interface is used.

jcoro replied:
Does the id parameter require a `@Nullable` annotation?
Putting the javadoc in the interface makes sense.  Along those lines, should I move the javadoc for each service we have to each interface we have?

jdamerow replied:
> Does the id parameter require a @Nullable annotation?

The javadoc should just say that id can be null and what happens if that is the case. If I do an autocomplete on fileSystemService, trying to choose a method I want to know if the method does what I need (e.g. just give me the path to the user directory if id is null). Does that make sense?

> Putting the javadoc in the interface makes sense. Along those lines, should I move the javadoc for each service we have to each interface we have?

Yes, ideally. I mention it here, since that was one of the method where the javadoc is really needed as I have several options how to call it.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
When I click on "My Datasets" I get a 404

jcoro commented:
> When I click on "My Datasets" I get a 404

It's likely a role issue.  I'm guessing you signed in as the default admin?  I thought that the default admin had a role of USER as well, but it may not.  Either way, I changed the relevant antMatcher to:
`.antMatchers("/auth/**").hasAnyRole("USER", "ADMIN")`
and it seems to work.  It should work currently if you sign in as a non-admin user.

jdamerow commented:
The menu option should either not show up if the user doesn't have permissions or it needs to show a unauthorized error page but not the default tomcat 404.

jcoro commented:
> The menu option should either not show up if the user doesn't have permissions or it needs to show a unauthorized error page but not the default tomcat 404.

Agreed - it was a mistake on my part, but are you saying that the above change/fix is insufficient (i.e., adding the ADMIN role to the antMatcher)? In other words, are you saying that an admin should not be able to have or access datasets in the application?

jdamerow commented:
> Agreed - it was a mistake on my part, but are you saying that the above change/fix is insufficient (i.e., adding the ADMIN role to the antMatcher)? In other words, are you saying that an admin should not be able to have or access datasets in the application?

Yes, if I login with another user that does not have user or admin privileges (I think there is a manager role?), or if I simply try to access that page without logging in, it shouldn't show a 404, it should show a forbidden page (not the default tomcat one).

jcoro committed:
[STC-13] Added error-handling page.

jcoro commented:
I attempted to implement a custom error page for the entire application, but, when I dug into it, there was a little bit more to it than what I anticipated.  I made some assumptions regarding implementation, so I wanted lay out what I did and what my thinking was in case there is a better solution.

Adding:

`.and().exceptionHandling().accessDeniedPage("/error")`

may work for 403 errors, but it doesn't seem to work for other errors such as a signed in user visiting a page that doesn't exist.

To handle these errors, I added:

```
<error-page>
    <location>/error</location>
</error-page>
```

To the `web.xml` file.

This seemed to display the custom error page, but I then noticed that the security tags that are used to display the navigation buttons (i.e., sign in, sign out, etc.) fail to work on error pages.  Apparently, `springSecurityFilterChain` will not work on requests that result in an error as is outlined here:

[https://stackoverflow.com/questions/16393309/spring-security-tags-do-not-work-in-error-pages](https://stackoverflow.com/questions/16393309/spring-security-tags-do-not-work-in-error-pages)

I added: 

```
<dispatcher>REQUEST</dispatcher>
<dispatcher>ERROR</dispatcher>
```

to `web.xml`.  I looked, but I couldn't find a way to set the dispatcher type in the Java-based Spring Security config.

I believe that `SimpleUsers` admin has an ADMIN role, but no USER role and no entry in the database.  With no entry in the database, the default `SimpleUsers` admin cannot CRUD Datasets (there's no database user to tie them to) but users who are *granted* the ADMIN role can CRUD Datasets (users who are granted the ADMIN role  have the role of USER as well, and have an entry in the database).  Therefore the user navigation (i.e., the link to "My Datasets") only shows for those with role of USER, regardless of whether said user also has roles of ADMIN or MANAGER.

In addition, with regard to the error page, it acts a little differently than outlined, and I wanted to make sure I didn't miss any use cases:

If a non-logged in person tries to access /auth/datasets/** it redirects to the login page (not the error page - I believe this a Spring Security default?).

If a USER tries to access /auth/datasets/** it allows the user to CRUD their own Datasets.

If the default ADMIN (without a USER role) tries to access /auth/datasets/** (manually - the link doesn't appear) the 403 error page displays.

If a granted ADMIN (with a USER role) tries to access /auth/datasets/** it allows granted admin to CRUD their own Datasets (I figured where the ADMIN role can be invoked and revoked, it might make sense for a granted ADMIN to create or maintain their Datasets).

If a user with a MANAGER role tries to access /auth/datasets/**, the manager also has a USER role, so it allows managers to CRUD their own Datasets (I figured where the MANAGER role can be invoked and revoked, this might make sense for the same reason as ADMIN above).

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
> I attempted to implement a custom error page for the entire application, but, when I dug into it, there was a little bit more to it than what I anticipated. I made some assumptions regarding implementation, so I wanted lay out what I did and what my thinking was in case there is a better solution.
> 
> Adding:
> 
> `.and().exceptionHandling().accessDeniedPage("/error")`
> 
> may work for 403 errors, but it doesn't seem to work for other errors such as a signed in user visiting a page that doesn't exist.

Yes, this is only for access denied pages (hence the method name `accessDeniedPage`). I would set this to `403` or something like that and create a page for it (controller + template) that is a specific access denied page.

> 
> To handle these errors, I added:
> 
> ```
> <error-page>
>     <location>/error</location>
> </error-page>
> ```
> 
> To the `web.xml` file.

This should work for every error that our app doesn't handle. 404 is a fairly special one as in most cases we can't do too much about it (if our app doesn't feel responsible for serving a page it tells Tomcat that). I would make that a generic error page (leave it the way you have it.)

> 
> This seemed to display the custom error page, but I then noticed that the security tags that are used to display the navigation buttons (i.e., sign in, sign out, etc.) fail to work on error pages. Apparently, `springSecurityFilterChain` will not work on requests that result in an error as is outlined here:
> 
> https://stackoverflow.com/questions/16393309/spring-security-tags-do-not-work-in-error-pages
> 
> I added:
> 
> ```
> <dispatcher>REQUEST</dispatcher>
> <dispatcher>ERROR</dispatcher>
> ```
> 
> to `web.xml`. I looked, but I couldn't find a way to set the dispatcher type in the Java-based Spring Security config.

I wouldn't even bother with that and just remove all spring security related links. Just a message that says that there was an error and then a link to go back to the home page. Also, the way it is set up right now, are you sure the tags don't work? It looks like you are sending the user to an error page that is coming from the app `/error` for which spring security should work.

However, this page should only be shown if there is an error our app doesn't handle. Most errors that would stem from thrown exceptions should be handled via a controller advice, exception handler, or directly in the controllers.

> 
> I believe that `SimpleUsers` admin has an ADMIN role, but no USER role and no entry in the database. With no entry in the database, the default `SimpleUsers` admin cannot CRUD Datasets (there's no database user to tie them to) but users who are _granted_ the ADMIN role can CRUD Datasets (users who are granted the ADMIN role have the role of USER as well, and have an entry in the database). Therefore the user navigation (i.e., the link to "My Datasets") only shows for those with role of USER, regardless of whether said user also has roles of ADMIN or MANAGER.

Not sure what you mean by that, but you should be able to make a new user a "USER" (that should actually be happening by default when you activate a user). 

> 
> In addition, with regard to the error page, it acts a little differently than outlined, and I wanted to make sure I didn't miss any use cases:
> 
> If a non-logged in person tries to access /auth/datasets/** it redirects to the login page (not the error page - I believe this a Spring Security default?).
> 
> If a USER tries to access /auth/datasets/** it allows the user to CRUD their own Datasets.
> 
> If the default ADMIN (without a USER role) tries to access /auth/datasets/** (manually - the link doesn't appear) the 403 error page displays.
> 
> If a granted ADMIN (with a USER role) tries to access /auth/datasets/** it allows granted admin to CRUD their own Datasets (I figured where the ADMIN role can be invoked and revoked, it might make sense for a granted ADMIN to create or maintain their Datasets).
> 
> If a user with a MANAGER role tries to access /auth/datasets/**, the manager also has a USER role, so it allows managers to CRUD their own Datasets (I figured where the MANAGER role can be invoked and revoked, this might make sense for the same reason as ADMIN above).

It shouldn't matter what other roles a user has. If they have the "USER" role, they can manage datasets, if not they can't.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/config/SecurityContext.java:
```diff
@@ -24,20 +24,22 @@ public void configure(WebSecurity web) throws Exception {
                 .deleteCookies("JSESSIONID")
                 .logoutSuccessUrl("/")
-                .and().exceptionHandling().accessDeniedPage("/403")
+                .and().exceptionHandling().accessDeniedPage("/error")
```
this should stay 403 and then add an additional error page for general erros

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/IDataFileService.java:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ * The interface associated with user-uploaded DataFiles.  The interface handles storing the
+ * files on the file system (using the FileSystemService) and storing DataFile data in the database.
```
The second sentence should go on the implementing class, since the interface by itself doesn't do anything. Here, javadoc should describe how or when to use this interface and what its implementations do (e.g. instances of this class do x). In theory, there could be another implementation that stores files in S3, so that this javadoc would not be correct anymore.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ErrorController.java:
```diff
@@ -0,0 +1,44 @@
+            }
+            default: {
+                errorMsg = "There was an Http error when accessing this resource";
+            }
```
instead of this huge switch statement, I would add a properties file that has all these messages (so that you just append the error code to a prefix, e.g. `error_msg_500`, and then get the message with that key). It's easier to maintain and doesn't require recompilation if a text changes.

jcoro committed:
[STC-13] Configured error-handling controllers and views.

diging-jenkins commented:
Congratulations, build succeeded!

jcoro commented:
I updated the error handling as indicated.  I think we were on the same page regarding roles and access to Datasets.  For clarification, I was just saying above that the default SimpleUsers admin (i.e., the admin with username: admin and password: admin) which is used to initially access the application and grant initial authorities doesn't seem to have a database entry associated with it, nor does it seem to have a USER role associated with it.  It is unique in that it has only an ADMIN role associated with it.  All subsequent users created within the application (whether they become ADMINs or MANAGER)s, have USER roles in addition to the other roles.

jcoro commented:
> > I attempted to implement a custom error page for the entire application, but, when I dug into it, there was a little bit more to it than what I anticipated. I made some assumptions regarding implementation, so I wanted lay out what I did and what my thinking was in case there is a better solution.
> > Adding:
> > `.and().exceptionHandling().accessDeniedPage("/error")`
> > may work for 403 errors, but it doesn't seem to work for other errors such as a signed in user visiting a page that doesn't exist.
> 
> Yes, this is only for access denied pages (hence the method name `accessDeniedPage`). I would set this to `403` or something like that and create a page for it (controller + template) that is a specific access denied page.
> 
> > To handle these errors, I added:
> > ```
> > <error-page>
> >     <location>/error</location>
> > </error-page>
> > ```
> > 
> > 
> > To the `web.xml` file.
> 
> This should work for every error that our app doesn't handle. 404 is a fairly special one as in most cases we can't do too much about it (if our app doesn't feel responsible for serving a page it tells Tomcat that). I would make that a generic error page (leave it the way you have it.)
> 
> > This seemed to display the custom error page, but I then noticed that the security tags that are used to display the navigation buttons (i.e., sign in, sign out, etc.) fail to work on error pages. Apparently, `springSecurityFilterChain` will not work on requests that result in an error as is outlined here:
> > https://stackoverflow.com/questions/16393309/spring-security-tags-do-not-work-in-error-pages
> > I added:
> > ```
> > <dispatcher>REQUEST</dispatcher>
> > <dispatcher>ERROR</dispatcher>
> > ```
> > 
> > 
> > to `web.xml`. I looked, but I couldn't find a way to set the dispatcher type in the Java-based Spring Security config.
> 
> I wouldn't even bother with that and just remove all spring security related links. Just a message that says that there was an error and then a link to go back to the home page. Also, the way it is set up right now, are you sure the tags don't work? It looks like you are sending the user to an error page that is coming from the app `/error` for which spring security should work.

> When I tested it, the Spring Security links didn't work.  I'm not sure, but I seem to remember investigating it and reading that the springSecurityFilterChain defaults to REQUEST requests and not ERROR requests in the filter mapping. 

> 
> However, this page should only be shown if there is an error our app doesn't handle. Most errors that would stem from thrown exceptions should be handled via a controller advice, exception handler, or directly in the controllers.
> 
> > I believe that `SimpleUsers` admin has an ADMIN role, but no USER role and no entry in the database. With no entry in the database, the default `SimpleUsers` admin cannot CRUD Datasets (there's no database user to tie them to) but users who are _granted_ the ADMIN role can CRUD Datasets (users who are granted the ADMIN role have the role of USER as well, and have an entry in the database). Therefore the user navigation (i.e., the link to "My Datasets") only shows for those with role of USER, regardless of whether said user also has roles of ADMIN or MANAGER.
> 
> Not sure what you mean by that, but you should be able to make a new user a "USER" (that should actually be happening by default when you activate a user).
> 
> > In addition, with regard to the error page, it acts a little differently than outlined, and I wanted to make sure I didn't miss any use cases:
> > If a non-logged in person tries to access /auth/datasets/** it redirects to the login page (not the error page - I believe this a Spring Security default?).
> > If a USER tries to access /auth/datasets/** it allows the user to CRUD their own Datasets.
> > If the default ADMIN (without a USER role) tries to access /auth/datasets/** (manually - the link doesn't appear) the 403 error page displays.
> > If a granted ADMIN (with a USER role) tries to access /auth/datasets/** it allows granted admin to CRUD their own Datasets (I figured where the ADMIN role can be invoked and revoked, it might make sense for a granted ADMIN to create or maintain their Datasets).
> > If a user with a MANAGER role tries to access /auth/datasets/**, the manager also has a USER role, so it allows managers to CRUD their own Datasets (I figured where the MANAGER role can be invoked and revoked, this might make sense for the same reason as ADMIN above).
> 
> It shouldn't matter what other roles a user has. If they have the "USER" role, they can manage datasets, if not they can't.

jdamerow commented on scriptoocloud/src/main/resources/error.properties:
```diff
@@ -0,0 +1,5 @@
+401 = 401. Unauthorized
+403 = 403. Access Denied
+404 = 404. Page Not Found
+500 = 500. Internal Server Error
```
Let's make these a little bit more user friendly, something like "400=The server was unable to understand your browser's request".

jdamerow reviewed (changes requested):
Besides that, the only UI comment I have that the renaming of a dataset is a little unintuitive. I expect it either to work like many other website, where it automatically saves after I made a change, or there is a save icon right next to the input field or something. Or that I have to click on a link like "rename dataset" and then I get an input field where I can make changes.

> For clarification, I was just saying above that the default SimpleUsers admin (i.e., the admin with username: admin and password: admin) which is used to initially access the application and grant initial authorities doesn't seem to have a database entry associated with it, nor does it seem to have a USER role associated with it.

Ah, yes, that is correct. It's a flaw that needs to be fixed in the plugin.

It's still missing the test cases, right? that's why it's a draft pull request?

jcoro committed:
[STC-13] Altered edit Dataset UI and added text to error properties.

jcoro commented:
I altered the edit Datasets UI to toggle the input field and Save button upon clicking "Rename this Dataset".  I also added to the error messages, and yes, the tests still have to be created.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/ErrorController.java:
```diff
@@ -0,0 +1,28 @@
+    @RequestMapping(value = "/error", method = RequestMethod.GET)
+    public String error(Model model, HttpServletResponse response) {
+        model.addAttribute("httpErrorMessage", "Http Error Code: " + environment.getProperty(
+                String.valueOf(response.getStatus()), "Unknown"));
```
I don't think that "Unknown" in your properties file? This would mean that this would throw an error if the status code is not in the properties file.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
For some reason, all files I upload get the name "pom.xml". It might be an issue with Maven though. Some variables are used by Maven (e.g. `project`) and using those as variable names in your templates can have surprising effects. I would rename the `file` variable and see if that fixes it.
![image](https://user-images.githubusercontent.com/8881141/108866443-364c3000-75c2-11eb-8508-206430b73e1c.png)

Besides these few things, looks good.

jcoro committed:
[STC-13] Altered file variable in auth/datasets/details.html and added Unknown to error.properties.

jcoro commented:
I changed the `file` variable in `auth/datasets/details.html` to `uploadedFile`.  Both versions work on my system, so I couldn't recreate the bug, nor could I find any info on Thymeleaf using Maven's variables so, I hope the fix works.

I added an Unknown key to the `error.properties` file.  I may be missing something in my understanding, however.  When I looked at the documentation, my interpretation was that the method would return the `String` "Unknown" if the key couldn't be found in `error.properties`  (javadoc below).  Having added Unknown to the `error.properties` file, however, I now see that my IDE resolves the property from the file when I hover over "Unknown" in the method call.  I'm curious just for the sake of improving my ability to interpret javadoc going forward.  If the method resolves the `defaultValue` property name, it seems like the documentation for `defaultValue` should be something like: "the default property name to resolve if the key value is not found".  

```
// org.springframework.core.env
// Interface PropertyResolver

String getProperty(String key,
                   String defaultValue)

// Return the property value associated with the given key, or defaultValue if the key cannot be resolved.

// Parameters
key - the property name to resolve
defaultValue - the default value to return if no value is found
```

Digging a little deeper, I believe that the method is implemented like:

```java
@Override
	public String getProperty(String key, String defaultValue) {
		String value = getProperty(key);
		return (value != null ? value : defaultValue);
	}
```

[https://github.com/spring-projects/spring-framework/blob/0d16c9100afc492a6f2258dfc2dfc9c9515438c2/spring-core/src/main/java/org/springframework/core/env/AbstractPropertyResolver.java#L168](https://github.com/spring-projects/spring-framework/blob/0d16c9100afc492a6f2258dfc2dfc9c9515438c2/spring-core/src/main/java/org/springframework/core/env/AbstractPropertyResolver.java#L168)

Again, it should work either way, now. I'm curious for the sake of understanding.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
No, you're right, I was confused. The default value is returned if the key is not found, no message for `Unknown` needed.

jdamerow commented:
The filename issue is resolved now. So, just reopen when you're done with the tests.

jcoro committed:
[STC-13] Added Testing for FileSystemService, DatasetService, DataFileService.

jcoro commented:
Added tests for FileSystemService, DatasetService, and DataFileService.

diging-jenkins commented:
Build failed. Maybe you broke the test cases?

jcoro commented:
All of the tests seem to pass but it looks like its failing some Jenkins checks - I'll look into it

jcoro committed:
[STC-13] Altered FileSystemServiceTest to test Jenkins build.

diging-jenkins commented:
Congratulations, build succeeded!

jcoro commented:
The issue breaking the build seems to have been in a fileSystemService test case - I'll refactor it.

jcoro committed:
[STC-13] Refactored FileSystemServiceTest to test Jenkins build.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+    @Test
+    public void test_createFile() throws FileSystemStorageException {
+
+        Dataset dataset = Mockito.mock(Dataset.class);
```
since dataset is just a domain object and doesn't contain any business logic, I wouldn't mock it. You can just set the properties you need.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+        Dataset dataset = Mockito.mock(Dataset.class);
+        Mockito.when(datasetRepository.findById(datasetId))
+                .thenReturn(Optional.of(dataset));
+        Mockito.when(dataset.addFile(dataFileRepository.save(Mockito.any()))).thenReturn(1L);
```
you only want 1L to be returned when a specific dataset is being saved, don't you? Not any?

jcoro replied:
> you only want 1L to be returned when a specific dataset is being saved, don't you? Not any?

I'm having some general trouble with testing, and this test is a good example.
How would I best go about not using `Mockito.any()` above?
I looked into `ArgumentCaptor` and `ArgumentMatcher`, but I'm not sure if these could be of use.

I could return the `dataFile` created in the test method, and then use its Id:

```
Mockito.when(dataFileRepository.save(Mockito.any())).thenReturn(dataFile);
Mockito.when(dataset.addFile(dataFileRepository.save(dataFile))).thenReturn(dataFile.getId());
```

This works, but it still uses `Mockito.any()`, so it doesn't seem to be a viable solution.

Similarly, when running this test, `dataset.files` is null so both `dataset.add(file)` and `dataFileRepository.save(file)` throw an NPE.  
I tried reflection to add a Hibernate `PersistentSet` to the `dataset`, but received:
`org.hibernate.LazyInitializationException: failed to lazily initialize a collection, could not initialize proxy - no Session`
So, I then annotated the test class with `@Transactional`, and received the error:
`Failed to retrieve PlatformTransactionManager for @Transactional test...`
So, then I then added the `PersistenceConfig` class to the `@ContextConfiguration` annotation, but received an error that Spring couldn't create an `EntityManagerFactory`.

I added a `HashSet` instead of a Hibernate `PersistenceSet` for the files attribute on the `dataset` but this seems like a clunky solution.

Generally, these sorts of problems were why I was previously using mocks and spies, etc. but I could use some guidance as to what the best approach is for these sorts of tests.

jdamerow replied:
> How would I best go about not using Mockito.any() above?

Here is an example. 
https://github.com/diging/citesphere/blob/b0fec7c0eea580b3f8f1f8bfa09ef3d81b8008c2/citesphere/src/test/java/edu/asu/diging/citesphere/core/util/model/impl/CitationHelperTest.java#L205
I would create an argument matcher that is initialized for example with filename, type, and dataset id and checks these values agains the passed datafile object. If they match, it returns the id.

For the Nullpointer exception, this might actually a hint that there is a bug in your code or at least the potential for a bug. When is `files` initialized? I couldn't find it in either the dataset class, nor the datafile service or the dataset service. So, given that your test is throwing a NPE there could potentially a scenario that your code would throw that at runtime as well. I'm not sure, what Hibernate exactly does, but maybe it automatically adds a persistent set to any one to many relationship, but a) I wouldn't trust that this is always happening or that another framework would do the same (in case we ever exchange the ORM) and b) if at any point the code is changed in a way that a dataset does not have to be persisted yet to add a datafile, then this would also fail. 

Does that make sense?

jcoro replied:
Thanks - I think that makes sense.  So, if I implement checking for a null file Set in the `Dataset` class (where the NPE was coming from) what implementation of `Set` should I use?  Presumably, Hibernate's `PersistentSet` is Hibernate-specific, but I see from the docs that the underlying collection is a `HashSet`.  Could I do:
```
@Override
    public Long addFile(DataFile dataFile) {
        if (files == null){
            files = new HashSet<>();
        }
        files.add(dataFile);
        return dataFile.getId();
    }
```

jdamerow replied:
Yes, I would just use a HashSet. Hibernate will replace that with a PersistentSet when you retrieve an object from the DB.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+        Mockito.verify(datasetRepository).save(dataset);
+        Mockito.verify(fileSystemService).createFileInDirectory(username, type, datasetIdString,
+                indexBasedFilename, bytes);
+    }
```
Test cases needed for failed file creation.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+        PageRequest pageRequest = PageRequest.of(pageNumber + 1, pageSize);
+        PageRequest spyPageRequest = Mockito.spy(pageRequest);
+        for (int i = 0; i < pageSize + numberInSecondPage; i++) {
+            DataFile dataFile = Mockito.mock(DataFile.class);
```
same as before, domain classes usually don't need to be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,110 @@
+        List<DataFile> files = new ArrayList<>();
+        List<DataFile> spyFiles = Mockito.spy(files);
+        PageRequest pageRequest = PageRequest.of(pageNumber + 1, pageSize);
+        PageRequest spyPageRequest = Mockito.spy(pageRequest);
```
What do you need a spy for?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+    private Dataset dataset;
+
+    @Spy
+    private SimpleUser simpleUser;
```
why do you need spies? domain objects shouldn't need that in most cases, and the filesystemservice should be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+    public void test_createDataset() throws FileSystemStorageException {
+        IUser user = simpleUser;
+        Mockito.when(datasetRepository.save(Mockito.any())).thenReturn(dataset);
+        Mockito.when(dataset.getId()).thenReturn(1L);
```
just set the id and use the real object

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Mockito.when(dataset.getId()).thenReturn(1L);
+        Assertions.assertDoesNotThrow(() -> datasetService.createDataset(username, user));
+        Mockito.verify(fileSystemService).addDirectories(username, type, "1");
+    }
```
same as before, needs test cases for failing scenarios.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+    @Test
+    public void test_editDataset() {
+        String newName = "newName";
+        Mockito.when(datasetRepository.findById(Mockito.any())).thenReturn(Optional.of(dataset));
```
you wanna find a dataset with a specific id, don't you? not just any.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Mockito.when(datasetRepository.findById(Mockito.any())).thenReturn(Optional.of(dataset));
+        Assertions.assertDoesNotThrow(() -> datasetService.editDataset(dataset.getId(), newName));
+        Mockito.verify(datasetRepository).save(dataset);
+    }
```
same as before. are there failing scenarios? they need test cases.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Mockito.when(datasetRepository.findById(Mockito.any())).thenReturn(optional);
+        Assertions.assertDoesNotThrow(() -> datasetService.deleteDataset(datasetId, username));
+        Mockito.verify(dataFileRepository).deleteAllByDatasetId(datasetId);
+        Mockito.verify(datasetRepository).deleteById(datasetId);
```
test cases for failing scenarios.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Optional<Dataset> optional = Optional.of(dataset);
+        Long datasetId = optional.get().getId();
+        Mockito.when(datasetRepository.findById(1L)).thenReturn(optional);
+        Assertions.assertDoesNotThrow(() -> datasetService.findById(datasetId));
```
what would be thrown here?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        }
+        Page<Dataset> datasetListSecondPage = new PageImpl<>(spyDatasets.subList(pageSize - 1,
+                (pageSize - 1 + numberInSecondPage)));
+        Page<Dataset> spyDatasetList = Mockito.spy(datasetListSecondPage);
```
what do you need the spies for?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        List<Dataset> spyDatasets = Mockito.spy(datasets);
+        Page<Dataset> datasetList = new PageImpl<>(spyDatasets);
+        PageRequest pageRequest = PageRequest.of(pageNumber, pageSize);
+        PageRequest spyPageRequest = Mockito.spy(pageRequest);
```
what do you need the spy for?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        String pathString = "pathString";
+
+        Mockito.when(dataFileRepository.findById(dataFileId)).thenReturn(Optional.of(dataFile));
+        Mockito.when(dataFile.getName()).thenReturn(fileName);
```
datafile does not need to be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Mockito.when(dataFile.getName()).thenReturn(fileName);
+        Mockito.when(datasetRepository.findById(datasetId)).thenReturn(Optional.of(dataset));
+        Mockito.when(dataFileRepository.findById(dataFile.getId())).thenReturn(Optional.of(dataFile));
+        Mockito.when(dataset.getFiles()).thenReturn(files);
```
dataset does not need to be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Set<DataFile> files = new HashSet<>();
+        files.add(dataFile);
+        File fileToBeDeleted = Mockito.mock(File.class);
+        Path path = Mockito.spy(Path.class);
```
what do you need a spy for?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,188 @@
+        Mockito.when(path.toString()).thenReturn(pathString);
+        Mockito.when(fileSystemService.deleteDirectoryOrFile(fileToBeDeleted)).thenReturn(true);
+
+        Assertions.assertDoesNotThrow(() -> datasetService.deleteFileFromDataset(datasetId,
```
what about test cases for when this does throw an exception?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,114 @@
-    
-    @BeforeEach
-    public void init() {
-        MockitoAnnotations.initMocks(this);   
```
why are those being removed?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,114 @@
+    private final String id = "1";
+
+    @Mock
+    private final FileSystemService fileSystemService;
```
since this is the test class for FileSystemService, it should not be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,114 @@
+    public void test_addDirectories() {
+        folder = folder.resolve(username + "/" + type + "/" + id);
+        FileSystemService spyFileSystemService = Mockito.spy(fileSystemService);
+        Mockito.when(spyFileSystemService.createPath(username, type, id)).thenReturn(folder);
```
`fileSystemService` is the class being tested, it should not be mocked.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,114 @@
-        Assertions.assertEquals(false, file.exists());
-    }   
-    
+    public void test_CreatePath_nullId() {
```
`CreatePath` should be lowercased

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,114 @@
+        String segmentType = segments[segments.length - 2];
+        Assertions.assertTrue(file.createNewFile());
+        FileSystemService spyFileSystemService = Mockito.spy(fileSystemService);
+        Mockito.when(spyFileSystemService.createPath(segmentUserName, segmentType, id)).thenReturn(file.toPath());
```
see above; should not be mocked.

jcoro committed:
[STC-13] Updated Dataset and DataFile tests.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/model/impl/Dataset.java:
```diff
@@ -0,0 +1,116 @@
+        }
+        if (dataFile == null){
+            dataFile = new DataFile();
+        }
```
if null is passed, it doesn't really make sense to add a new, empty datafile? Just return without adding it?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+
+    private static final byte[] BYTES = "abc".getBytes();
+
+    private static final OffsetDateTime CREATED_AT = OffsetDateTime.now();
```
the above constants do not need to be static.

jcoro replied:
Out of curiosity, is this because there is only going to be one instance of each test class, so there is no benefit to static constants?  Are there any downsides to making constants static?  Again, just curious.

jdamerow replied:
There is probably no downside of it, its just not very consistent since all the other variables are non-static. And yes, there will be just the one instance so what's the point. Generally, I would avoid static members in classes unless they really serve a purpose. They either easily introduce bugs (if they are not final and someone forgets that it's a shared variable) or they kind of hard-code something that can't be changed programmatically later. E.g. you have a class that has a static final property `maxUsers` and you set that to say 20. If the requirements change and now it should be possible to dynamically change that number, you have to rewrite your code. Does that make sense?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void test_createFile() throws FileSystemStorageException {
```
The name should have a third part describing the outcome (e.g. `_success`).

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void test_createFile_failed_data_file_exception() {
```
third part should be camelcase too (`failedData...`)

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void test_createFile_failed_file_system_exception() throws FileSystemStorageException {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void test_findFiles() {
```
same as above, needs third part.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+        PageRequest pageRequest = PageRequest.of(PAGE_NUMBER, PAGE_SIZE);
+        Mockito.when(dataFileRepository.findAllByDatasetId(DATASET_ID,
+                pageRequest)).thenReturn(fileList);
+        Assertions.assertEquals(fileList, dataFileService.findFiles(pageRequest, DATASET_ID));
```
I would test that the files list in `fileList` is empty to make sure the service doesn't add any files after retrieving it from the `dataFileRepository`.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+        PageRequest pageRequest = PageRequest.of(PAGE_NUMBER, PAGE_SIZE);
+        Mockito.when(dataFileRepository.findAllByDatasetId(DATASET_ID,
+                pageRequest)).thenReturn(fileList);
+        Assertions.assertEquals(fileList, dataFileService.findFiles(pageRequest, DATASET_ID));
```
I would also check the length of the list of files to make sure it is correctly passed from repository to caller.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+        Assertions.assertEquals(SECOND_PAGE_SIZE, dataFileListSecondPage.getTotalElements());
+    }
+
+    static class DataFileArgMatcher extends ArgumentMatcher<IDataFile> {
```
is there a reason that this is a static class?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_createDataset() throws FileSystemStorageException {
```
see previous comments

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    public void test_createDataset() throws FileSystemStorageException {
+        Mockito.when(datasetRepository.save(((Dataset) Mockito.argThat(
+                new DatasetServiceTest.DatasetArgMatcher(dataset))))).thenReturn((Dataset) dataset);
+        Assertions.assertDoesNotThrow(() -> datasetService.createDataset(DATASET_NAME, user));
```
you also want to make sure you get back a dataset object with the right properties set.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_editDataset() {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+                .thenReturn(Optional.of((Dataset) dataset));
+        Assertions.assertDoesNotThrow(() -> datasetService.editDataset(dataset.getId(), newName));
+        dataset.setName(newName);
+        Mockito.when(datasetRepository.save((Dataset) dataset)).thenReturn(newDataset);
```
this should go before you method call or I don't think it does much. Also, since the return value from the save method is never used, you actually don't need this at all. Instead remove line 156 since your code should do that and then test that `dataset.getName` returns `newName`.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+        Assertions.assertDoesNotThrow(() -> datasetService.editDataset(dataset.getId(), newName));
+        // should not be called this time - dataset not present
+        Mockito.verify(datasetRepository, Mockito.never()).save((Dataset) dataset);
+        Assertions.assertNotEquals(newDataset.getName(), dataset.getName());
```
newDataset is never used is it, that's confusing. If the datasetId doesn't exist nothing should happen right? So just the verify should be enough.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_deleteDataset() {
```
see above.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_findById() {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    public void test_findById() {
+        Mockito.when(datasetRepository.findById(DATASET_ID)).thenReturn(
+                Optional.of((Dataset) dataset));
+        Assertions.assertEquals(dataset, Optional.of((Dataset) dataset).get());
```
datasetservice is not called in this test.

jdamerow replied:
right now you're testing the `get` method of `Optional`.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_findDatasets() {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+        Page<Dataset> datasetList = new PageImpl<>(datasetPage);
+        PageRequest pageRequest = PageRequest.of(PAGE_NUMBER, PAGE_SIZE);
+        Mockito.when(datasetRepository.findAllByUser(user, pageRequest)).thenReturn(datasetList);
+        Assertions.assertEquals(datasetList, datasetService.findDatasets(pageRequest, user));
```
I would check the length of the datasetList. Since you're comparing objects not the content, if the service methods modifies datasetList after retrieval from the repository (e.g. adding or removing elements) then your test would pass although it should fail.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_findAll() {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+        Page<Dataset> datasetList = new PageImpl<>(datasets);
+        PageRequest pageRequest = PageRequest.of(PAGE_NUMBER, PAGE_SIZE);
+        Mockito.when(datasetRepository.findAll(pageRequest)).thenReturn(datasetList);
+        Assertions.assertEquals(datasetList, datasetService.findAll(pageRequest));
```
same as above pretty much. That is probably true for all these test methods.

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_deleteFileFromDataset() throws FileSystemStorageException {
```
see above

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+                .thenReturn(Optional.of((DataFile) dataFile));
+        Mockito.when(fileSystemService.createPath(USERNAME, TYPE,
+                DATASET_ID_STRING)).thenReturn(path);
+        Mockito.when(fileSystemService.createPath(USERNAME, TYPE, DATASET_ID_STRING))
```
Why is that here twice?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+                DATA_FILE_ID));
+
+        Mockito.verify(datasetRepository).save((Dataset) Mockito
+                .argThat(new DatasetServiceTest.DatasetArgMatcher((dataset))));
```
you also want to make sure that the datafile has been deleted from the files list, don't you?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    @Test
+    public void test_deleteFileFromDataset_no_file_on_server() throws FileSystemStorageException {
```
see before (camelcase missing)

jdamerow replied:
same is true for the previous methods

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetServiceTest.java:
```diff
@@ -0,0 +1,326 @@
+        Mockito.verify(dataFileRepository).deleteById(DATA_FILE_ID);
+    }
+
+    static class DatasetArgMatcher extends ArgumentMatcher<IDataset> {
```
does this class need to be static? Also, is that the same as in the other test class? If so, just make it a standalone class and reuse.

jcoro replied:
It seems that the `ArgumentMatchers` require a new instance of the inner class to work, but I was under the impression that it's good practice to make inner classes like this static.

I searched and found the following:

https://stackoverflow.com/questions/16147405/whats-the-advantage-of-making-an-inner-class-as-static-with-java

I also get a warning from IntelliJ when removing static.

There's one `ArgumentMatcher` for `Datasets` and one `ArgumentMatcher` for `DataFiles` in their respective test classes.

jdamerow replied:
A static inner class can be used outside of the enclosing class without the need for an object of the enclosing class. Since no other class should be using the argument matcher outside of the test class, there is no need to make it static. If another class would use it, I would move it to its own file. 

Inner classes are good for logical grouping of classes, e.g. MapEntry only makes sense in the context of a Map, so it makes sense to have it as an inner class. Since the test is the only class using the matcher class, there is no need to make it available outside and making it static kind of sends the message that it can be used by other classes. But it would actually make it less straight forward to maintain if someone else would reuse it for their test cases in another class. Does this make sense? Maybe making it private will satisfy IntelliJ?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,128 @@
-        Assertions.assertEquals(false, file.exists());
-    }   
-    
+    public void test_addDirectories() throws FileSystemStorageException, IOException {
```
see previous comments

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,128 @@
-    }   
-    
+    public void test_addDirectories_failed() {
+        Assertions.assertThrows(NullPointerException.class,
```
when is a nullpointer thrown? That is usually an indicator that something should be fixed (as nullpointers are in most cases easily avoidable)

jcoro replied:
The `NullPointerExceptions` for both test cases are thrown in the `FileSystemService` `createPath()` method - in particular, Java's `Paths.get(null)` therein throws a `NullPointerException`.  I couldn't quite figure out how to test failure for `addDirectories()` and similarly, failure for `createPath()`.  Every combination of strings I tried is cleaned in the `createPath()` method and returns successfully.  I remember you mentioned a while back that it is generally bad practice to catch `NullPointerException`s, so I'm wondering if and how `createPath()` should be altered to deal with null inputs, and/or how to test the method(s) appropriately?

jdamerow replied:
I think I would have addDirectories return a boolean (true if directory was created, otherwise false). Then check in the method if the passed values are all non null. If username or type are null, return false, which is the condition your test can check (plus if the folder was created maybe). similar for the `createPath` method. I would actually probably allow to just pass in `username` and the other values null to just create a user folder. And then check if at least username is non null. If it is null, return null which your test can check for. Does this make sense?

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,128 @@
     @Test
-    public void test_deleteDirectoryOrFile_noSuchFile() throws IOException {
-        File file = new File("NotOnDrive");
+    public void test_createPath() throws FileSystemStorageException {
```
see above comments

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,128 @@
+
+    @Test
+    public void test_createPath_failed() {
+        Assertions.assertThrows(NullPointerException.class,
```
see above comment about null pointer exception

jdamerow commented on scriptoocloud/src/test/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemServiceTest.java:
```diff
@@ -1,72 +1,128 @@
+    }
+
+    @Test
+    public void test_deleteDirectories() throws FileSystemStorageException, IOException {
```
see previous comments

jdamerow commented:
After re-reading the ticket, I just realized that I think this is not yet addressed:
"Users should be able to create different version of datasets. When a user uploads a new dataset, the uploaded dataset would be version 1 of that dataset."

jcoro commented:
> After re-reading the ticket, I just realized that I think this is not yet addressed:
> "Users should be able to create different version of datasets. When a user uploads a new dataset, the uploaded dataset would be version 1 of that dataset."

I'm working on this now - would the version be a Long or a String?

jdamerow commented:
I would probably make it an int or a long to make it easier to compare.

jcoro committed:
[STC-13] Added version, description, and createdDate to Datasets and altered storage paths accordingly.

jcoro commented:
I implemented the versioning for Datasets, setting the default version for new Datasets to 1.  As for editing the Datasets (i.e., changing the version number) can you give details on how this would affect 1. the Dataset in the database (would this create a new database entry or simply alter the version field on an existing entry?  If it's the latter, what record is kept of previous versions if any?) and 2. the Dataset on the file system (would version 1 files be copied over to the version 2 directory?).

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Followup versions of a dataset would either be uploaded by a user (separate ticket) or if a script alters a dataset then script outputs could be come new versions of a dataset. The version wouldn't be increased by changing the existing dataset. It would be a conscious action of the user (e.g. the user cleaned their original dataset more and uploads a cleaned version or uses a script to do something with a dataset and the output of the script would become a new version). Does that help?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetService.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    public Dataset createDataset(String name, IUser user, String version, String description) throws DatasetStorageException {
```
why is this method expecting a string for version instead of a Long?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -27,89 +29,88 @@
+                path = Paths.get(getRootLocationString(), username);
+            } else if (username != null && (type != null && id != null && version != null)) {
+                path = Paths.get(getRootLocationString(), username, type, id, version);
+            } else if (username != null && (type != null && id != null)) {
```
you don't need `()` around `&&` statements; `true && (false && true)` is the same as `true && false && true`.

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -27,89 +29,88 @@
+        }
+        if ((type == null || id == null || version == null) &&
+                (type != null || id != null || version != null)) {
+            return false;
```
If I'm not mistaken `createPath` will just return null if these conditions are not met? Why not just check for null after line 50?

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/forms/DatasetForm.java:
```diff
@@ -0,0 +1,52 @@
+
+import javax.validation.constraints.NotEmpty;
+
+public class DatasetForm {
```
what's the different between this and `DatasetEditForm`? You can probably use the same class.

jcoro committed:
[STC-13] Altered createDataset method and deleted EditDatasetForm.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * The Service class associated with user-uploaded DataFiles.  The class handles storing the
+ * files on the file system (using the FileSystemService) and storing DataFile data in the database.
+ */
```
needs author information

jdamerow commented on scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/FileSystemService.java:
```diff
@@ -27,89 +29,87 @@
+        } else {
+            return "";
+        }
+    }
```
This wouldn't change the logic but the tertiary operator woudl work well here
```
return rootLocationString != null ? rootLocationString : "";
```

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Adding a file to an existing dataset (existing before this change) doesn't work. I assume it's because your datafileService create method assumes that there is a version (but for old datasets the version would be null).

Also, is there a reason this is still a draft PR?

jcoro commented:
> Adding a file to an existing dataset (existing before this change) doesn't work. I assume it's because your datafileService create method assumes that there is a version (but for old datasets the version would be null).
> 
> Also, is there a reason this is still a draft PR?


Is adding a `DataFile` causing a NPE? In testing, I deleted my test datasets and started fresh so I'm trying to piece together the error without being able to recreate it.

A null `Dataset` version would throw an NPE here in the `DataFileService` `createFile` method:

`version = dataset.get().getVersion().toString();`

which I can replace with:

`version = dataset.get().getVersion() != null ? dataset.get().getVersion().toString() : null;`

I also created the following additional test:

```
    @Test
    public void test_createFile_datasetWithNullVersion() throws FileSystemStorageException {
        dataset.setVersion(null);
        dataFile.setDataset((Dataset) dataset);
        Mockito.when(datasetRepository.findById(DATASET_ID))
                .thenReturn(Optional.of((Dataset) dataset));
        Mockito.when(dataFileRepository.save((DataFile) Mockito.argThat(
                new DataFileArgMatcher(dataFile)))).thenReturn((DataFile) dataFile);
        Assertions.assertDoesNotThrow(() -> dataFileService.createFile(BYTES, DATASET_ID_STRING,
                USERNAME, FILE_NAME, FILE_EXTENSION));
        Mockito.verify(fileSystemService).createFileInDirectory(USERNAME, FILE_TYPE, DATASET_ID_STRING,
                null, INDEX_BASED_FILENAME, BYTES);
    }
```

which passes.

As for the creation of the correct path on the file system, I suspect that the `createFileInDirectory` method in `FileSystemService` should allow for a null `Dataset` version, as `Paths.get` throws a NPE if any of its String arguments are null.

Does this sound like it would solve the particular issue you experienced?

As for the draft pull request, you mentioned a while back to reopen the pull request so that comments remain visible. I've continued to do that since then.  Should I open a new (non-draft) pull request for the next commit?

jdamerow commented:
I haven't checked the logs what the specific exception was, but I suspect that `getVersion().toString()` was the culprit. If fixing this would still not work because of the `createFileInDirectory` method that you need to fix that too (e.g. set a default value).

Regarding the draft pull request, you should be able to convert this PR into a regular PR. There is a button somewhere I believe.

jcoro committed:
[STC-13] Allowed for DataFile upload when version of Dataset is null.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Deleting a file from a dataset with no version has the same issue with null versions.

jcoro committed:
[STC-13] Allowed for DataFile deletion when version of Dataset is null.

jcoro commented:
Okay, I believe this should now be fixed.  It shouldn't throw an NPE if version is null and the `createPath` method already allows for a null version parameter.  Null versions shouldn't be an issue for editing `Datasets` - at this point, only the name can be edited.

diging-jenkins commented:
Congratulations, build succeeded!
